chore: document sdk helpers (#774)
defenseunicorns/pepr: docs/030_user-guide/015_sdk.md

Reviewer: All the type-casting removal stuff aside, do we want to add types to the examples for reals?  Stuff like:

```
// function getOwnerRefFrom(customResource: GenericKind) {}
const ownerRef = getOwnerRefFrom(customResource);
```

The upside would be that the appropriate types were a little easier to grok.

The downside would be that stuff like this is probably better done via automation / scraping -- maybe better to wait on that until we can just do proper doc-gen instead of hand-rolling them?

Author: Added to the @param tags.

Reviewer: Or, perhaps we could just sneak the types into the @param tags?  Something akin to this:

```
Returns the owner reference for a custom resource. Accepts the following parameters:

- **@param customResource: GenericKind** The custom resource to get the owner reference for

[...]

const ownerRef = getOwnerRefFrom(customResource);
```

Author: Updated

# Pepr SDK

> See also [Helpers](./030_actions/010_mutate.md) for information on helpers for mutate actions.

Reviewer: Probably this should be moved toward the end-ish..?

It's nice that this link in the doc, for sure, but it's a little strange that the first thing I see on the Pepr SDK page isn't related to Pepr SDK, methinks.  :thinking:

Author: Now moved to the end.

## `containers`

Returns list of all containers in a pod. Accepts the following parameters:

- **@param request** The request/pod to get the containers from
- **@param containerType** The type of container to get

**Usage:**

**_Get all containers_**

```typescript
let result = containers(peprValidationRequest)
```

**_Get only the standard containers_**

```typescript
let result = containers(peprValidationRequest, "containers")
```

**_Get only the init containers_**

```typescript
let result = containers(peprValidationRequest, "initContainers")
```

**_Get only the ephemeral containers_**

```typescript
let result = containers(peprValidationRequest, "ephemeralContainers")
```

---

## `getOwnerRefFrom`

Returns the owner reference for a custom resource. Accepts the following parameters:

- **@param cr** The custom resource to get the owner reference for

**Usage:**

```typescript

> const ownerRef = getOwnerRefFrom(cr as GenericKind);

Reviewer: ```suggestion
const ownerRef = getOwnerRefFrom(customResource);
```

Seems better -- for docs purposes -- to just assume that the given param is going to be of the right type rather than build-in a typecast, IMO.

```

---

## `writeEvent`

Write a K8s event for a CRD. Accepts the following parameters:

- **@param cr** The custom resource to write the event for
- **@param event** The event to write, should contain a human-readable message for the event
- **@param eventType** The type of event to write, for example "Warning"
- **@param eventReason** The reason for the event, for example "ReconciliationFailed"
- **@param reportingComponent** The component that is reporting the event, for example "uds.dev/operator"
- **@param reportingInstance** The instance of the component that is reporting the event, for example process.env.HOSTNAME

**Usage:**

```typescript
writeEvent(

>   cr as GenericKind,

Reviewer: ```suggestion
  customResource,
```

Same reasoning as above -- don't think that showing off a typecast is really necessary / an ideal look.

  event,
  "Warning",
  "ReconciliationFailed",
  "uds.dev/operator",

>   process.env.HOSTNAME as string,

Reviewer: ```suggestion
  process.env.HOSTNAME,
```

Same reasoning as above.

# Pepr SDK

## `containers`

Returns list of all containers in a pod. Accepts the following parameters:

- **@param peprValidationRequest** The request/pod to get the containers from
- **@param containerType** The type of container to get

**Usage:**

**_Get all containers_**

```typescript
let result = containers(peprValidationRequest)
```

**_Get only the standard containers_**

```typescript
let result = containers(peprValidationRequest, "containers")
```

**_Get only the init containers_**

```typescript
let result = containers(peprValidationRequest, "initContainers")
```

**_Get only the ephemeral containers_**

```typescript
let result = containers(peprValidationRequest, "ephemeralContainers")
```

---

## `getOwnerRefFrom`

Returns the owner reference for a Kubernetes resource. Accepts the following parameters:

- **@param kubernetesResource: GenericKind** The Kubernetes resource to get the owner reference for

**Usage:**

```typescript
const ownerRef = getOwnerRefFrom(kubernetesResource);

> 

Reviewer: I think you're  missing the backticks to close the typescript block here.  :thinking:

---

## `writeEvent`

Write a K8s event for a CRD. Accepts the following parameters:

- **@param kubernetesResource: GenericKind** The Kubernetes resource to write the event for
- **@param event** The event to write, should contain a human-readable message for the event
- **@param eventType** The type of event to write, for example "Warning"
- **@param eventReason** The reason for the event, for example "ReconciliationFailed"
- **@param reportingComponent** The component that is reporting the event, for example "uds.dev/operator"
- **@param reportingInstance** The instance of the component that is reporting the event, for example process.env.HOSTNAME

**Usage:**

```typescript
writeEvent(
  kubernetesResource,
  event,
  "Warning",
  "ReconciliationFailed",
  "uds.dev/operator",
  process.env.HOSTNAME,
);
```

---

## `sanitizeResourceName`

Returns a sanitized resource name to make the given name a valid Kubernetes resource name. Accepts the following parameter:

- **@param resourceName** The name of the resource to sanitize

**Usage:**

```typescript
const sanitizedResourceName = sanitizeResourceName(resourceName)
```

---

## See Also

> Looking for information on the Pepr mutate helpers? See [Helpers](./030_actions/010_mutate.md) for information on helpers for mutate actions.

Reviewer: ```suggestion
Looking for information on the Pepr mutate helpers? See [Mutate Helpers](./030_actions/010_mutate.md#mutate-helpers) for information on those.
```

I'm not 100% on the link-anchor syntax -- worth testing, for sure -- but I think something like this might work better cuz it takes a reader directly to the appropriate section of the target page, no?

# Pepr SDK

## `containers`

Returns list of all containers in a pod. Accepts the following parameters:

- **@param peprValidationRequest** The request/pod to get the containers from
- **@param containerType** The type of container to get

**Usage:**

**_Get all containers_**

```typescript
let result = containers(peprValidationRequest)
```

**_Get only the standard containers_**

```typescript
let result = containers(peprValidationRequest, "containers")
```

**_Get only the init containers_**

```typescript
let result = containers(peprValidationRequest, "initContainers")
```

**_Get only the ephemeral containers_**

```typescript
let result = containers(peprValidationRequest, "ephemeralContainers")
```

---

## `getOwnerRefFrom`

> Returns the owner reference for a custom resource. Accepts the following parameters:
> 
> - **@param customResource: GenericKind** The custom resource to get the owner reference for
> 
> **Usage:**
> 
> ```typescript
> const ownerRef = getOwnerRefFrom(customResource);
> ```

Reviewer 2: ```suggestion
Returns the owner reference for a Kubernetes resource. Accepts the following parameters:

- **@param kubernetesResource: GenericKind** The Kubernetes resource to get the owner reference for

**Usage:**

```typescript
const ownerRef = getOwnerRefFrom(kubernetesResource);
```
```

---

## `writeEvent`

Write a K8s event for a CRD. Accepts the following parameters:

> - **@param customResource: GenericKind** The custom resource to write the event for

Reviewer 2: ```suggestion
- **@param kubernetesResource: GenericKind** The Kubernetes resource to write the event for
```

- **@param event** The event to write, should contain a human-readable message for the event
- **@param eventType** The type of event to write, for example "Warning"
- **@param eventReason** The reason for the event, for example "ReconciliationFailed"
- **@param reportingComponent** The component that is reporting the event, for example "uds.dev/operator"
- **@param reportingInstance** The instance of the component that is reporting the event, for example process.env.HOSTNAME

**Usage:**

```typescript
writeEvent(

>   customResource,

Reviewer 2: ```suggestion
  kubernetesResource,
```
